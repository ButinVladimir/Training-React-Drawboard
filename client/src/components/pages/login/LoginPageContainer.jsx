import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import LoginPage from './LoginPage';
import { boardPage } from '../../../routes';

class LoginPageContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      locked: false,
      errorMessage: '',
      room: '',
    };

    this.onError = this.onError.bind(this);
    this.onSetRoom = this.onSetRoom.bind(this);
    this.onLogin = this.onLogin.bind(this);

    const { registerErrorHandlers } = this.props;
    registerErrorHandlers(this.onError);
  }

  onSetRoom(room) {
    const { locked } = this.state;

    if (locked) {
      return;
    }

    this.setState({ errorMessage: '', room });
  }

  onLogin() {
    const { locked, room } = this.state;
    const { onLogin } = this.props;

    if (locked) {
      return;
    }

    this.setState({ locked: true, errorMessage: '' });
    onLogin(room);
  }

  onError(error) {
    this.setState({ locked: false, errorMessage: error });
  }

  render() {
    const { isLoggedIn } = this.props;
    const { locked, errorMessage, room } = this.state;

    return (
      <Fragment>
        {isLoggedIn && (
          <Redirect to={boardPage} />
        )}
        {!isLoggedIn && (
          <LoginPage
            locked={locked}
            errorMessage={errorMessage}
            room={room}
            onSetRoom={this.onSetRoom}
            onLogin={this.onLogin}
          />
        )}
      </Fragment>
    );
  }
}

LoginPageContainer.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  registerErrorHandlers: PropTypes.func.isRequired,
  onLogin: PropTypes.func.isRequired,
};

export default LoginPageContainer;
