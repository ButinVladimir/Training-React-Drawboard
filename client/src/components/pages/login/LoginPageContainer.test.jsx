import React from 'react';
import { shallow, mount } from 'enzyme';
import LoginPageContainer from './LoginPageContainer';

describe('LoginPageContainer', () => {
  it('renders properly when user is not logged in', () => {
    const registerErrorHandlersMock = jest.fn();
    const component = shallow(<LoginPageContainer
      isLoggedIn={false}
      registerErrorHandlers={registerErrorHandlersMock}
      onLogin={() => {}}
    />);

    expect(component).toMatchSnapshot();
    expect(registerErrorHandlersMock).toHaveBeenCalledWith(component.instance().onError);
  });

  it('renders properly when user is logged in', () => {
    const registerErrorHandlersMock = jest.fn();
    const component = shallow(<LoginPageContainer
      isLoggedIn
      registerErrorHandlers={registerErrorHandlersMock}
      onLogin={() => {}}
    />);

    expect(component).toMatchSnapshot();
    expect(registerErrorHandlersMock).toHaveBeenCalledWith(component.instance().onError);
  });

  it('handles onSetRoom event when form is not locked', () => {
    const room = 'newRoom';
    const component = mount(<LoginPageContainer
      isLoggedIn={false}
      registerErrorHandlers={() => {}}
      onLogin={() => {}}
    />);

    component.find('input[name="room"]').simulate('change', { target: { value: room } });
    const componentState = component.state();

    expect(componentState).toMatchObject({
      errorMessage: '',
      room,
    });
  });

  it('handles onSetRoom event when form is locked', () => {
    const room = 'newRoom';
    const component = mount(<LoginPageContainer
      isLoggedIn={false}
      registerErrorHandlers={() => {}}
      onLogin={() => {}}
    />);

    component.setState({ locked: true });
    component.find('input[name="room"]').simulate('change', { target: { value: room } });
    const componentState = component.state();

    expect(componentState).not.toMatchObject({
      errorMessage: '',
      room,
    });
  });

  it('handles onLogin event when form is not locked', () => {
    const room = 'newRoom';
    const onLoginMock = jest.fn();
    const component = mount(<LoginPageContainer
      isLoggedIn={false}
      registerErrorHandlers={() => {}}
      onLogin={onLoginMock}
    />);

    component.setState({ room });
    component.find('button[name="enter"]').simulate('click');
    const componentState = component.state();

    expect(componentState).toMatchObject({
      locked: true,
      errorMessage: '',
    });
    expect(onLoginMock).toHaveBeenCalledWith(room);
  });

  it('handles onLogin event when form is locked', () => {
    const room = 'newRoom';
    const onLoginMock = jest.fn();
    const component = mount(<LoginPageContainer
      isLoggedIn={false}
      registerErrorHandlers={() => {}}
      onLogin={onLoginMock}
    />);

    component.setState({ locked: true, room });
    component.find('button[name="enter"]').simulate('click');
    const componentState = component.state();

    expect(componentState).toMatchObject({
      locked: true,
    });
    expect(onLoginMock).not.toHaveBeenCalled();
  });

  it('handles onError', () => {
    const errorMessage = 'An error occurred';
    let errorCb;
    const component = mount(<LoginPageContainer
      isLoggedIn={false}
      registerErrorHandlers={(cb) => { errorCb = cb; }}
      onLogin={() => {}}
    />);

    errorCb(errorMessage);
    const componentState = component.state();

    expect(componentState).toMatchObject({
      locked: false,
      errorMessage,
    });
  });
});
