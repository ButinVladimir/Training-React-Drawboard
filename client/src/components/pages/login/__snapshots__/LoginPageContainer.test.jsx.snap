// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPageContainer renders properly when user is logged in 1`] = `
<Fragment>
  <Redirect
    push={false}
    to="/board"
  />
</Fragment>
`;

exports[`LoginPageContainer renders properly when user is not logged in 1`] = `
<Fragment>
  <LoginPage
    errorMessage=""
    locked={false}
    onLogin={[Function]}
    onSetRoom={[Function]}
    room=""
  />
</Fragment>
`;
